libc/test/src/unistd/pipe2_test.cpp

```diff
@@ -0,0 +1,30 @@
+//===-- Unittests for pipe2 -----------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include "src/errno/libc_errno.h"
+#include "src/unistd/close.h"
+#include "src/unistd/pipe2.h"
+
+#include "test/UnitTest/ErrnoSetterMatcher.h"
+#include "test/UnitTest/Test.h"
+
+using namespace LIBC_NAMESPACE::testing::ErrnoSetterMatcher;
+
+TEST(LlvmLibcPipe2Test, SmokeTest) {
+  int pipefd[2];
+  ASSERT_THAT(LIBC_NAMESPACE::pipe2(pipefd, 0), Succeeds());
+  ASSERT_THAT(LIBC_NAMESPACE::close(pipefd[0]), Succeeds());
+  ASSERT_THAT(LIBC_NAMESPACE::close(pipefd[1]), Succeeds());
+}
+
+TEST(LlvmLibcPipe2ErrTest, SmokeTest) {
+  int pipefd[2];
+  ASSERT_THAT(LIBC_NAMESPACE::pipe2(pipefd, -1), Fails(EINVAL));
+  ASSERT_THAT(LIBC_NAMESPACE::pipe2(nullptr, 0), Fails(EFAULT));
+}
+
+// TODO: Functionality tests
```

Comment:
The tests aren't as thorough as https://github.com/llvm/llvm-project/pull/85514/files, but I think that's ok for syscall wrappers.

Reply:
syscall wrapper tests just need to exercise the libc side, so for most syscalls we just need a success and an error. It's also good to check that the values it returns are valid, since it's easy to forget to unpoison memory for sanitizers.

Reply:
The tests at https://github.com/llvm/llvm-project/pull/85514/files seem to be testing the kernel itself. For instance they verify that it's possible to read/write to the file descriptors, and that the descriptors are unique, etc. AFAIK this is outside the scope of libc. We assume the kernel is correct.

As far as the tests in the new PR go,
- We've verified that when return code is `-1`, `libc_errno` is populated.
- We've verified that `pipefd` holds valid file descriptors when `pipe2` returns 0. (If they were invalid [`close`](https://www.man7.org/linux/man-pages/man2/close.2.html) would not be successful.)

Is there anything I'm missing here?

Reply:
> Can it be removed here?

Sure. The linux kernel contains kselftests to exercise numerous things, including the syscall interfaces.  Ideally, the functional test suite we'd use is to build the Linux kernel's kselftests against llvm-libc, then run the result.